File: 01_CODE/app/src/main/java/com/beigel/leetSpeak_Generator/ProfileRepository.java
package com.beigel.leetSpeak_Generator;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ProfileRepository {
    private final ProfileManager profileManager;
    private final ExecutorService executorService;
    private final Handler mainHandler;

    public ProfileRepository(Context context) {
        this.profileManager = new ProfileManager(context);
        this.executorService = Executors.newSingleThreadExecutor();
        this.mainHandler = new Handler(Looper.getMainLooper());
    }

    public void loadFavoriteMode(FavoriteModeCallback callback) {
        executorService.execute(() -> {
            try {
                FavoriteModeResult result = loadFavoriteModeSync();
                mainHandler.post(() -> callback.onSuccess(result));
            } catch (Exception e) {
                mainHandler.post(() -> callback.onError(e));
            }
        });
    }

    private FavoriteModeResult loadFavoriteModeSync() {
        int favoriteMode = profileManager.getFavoriteMode();
        int favoriteCustomIndex = profileManager.getFavoriteCustomIndex();

        if (favoriteMode >= 0) {
            if (favoriteMode == ProfileManager.MODE_SIMPLE) {
                return FavoriteModeResult.simple();
            } else if (favoriteMode == ProfileManager.MODE_EXTENDED) {
                return FavoriteModeResult.extended();
            } else if (favoriteMode == ProfileManager.MODE_CUSTOM && favoriteCustomIndex >= 0) {
                if (profileManager.hasProfiles() && favoriteCustomIndex < profileManager.getProfiles().size()) {
                    CustomProfile profile = profileManager.getProfiles().get(favoriteCustomIndex);
                    return FavoriteModeResult.custom(favoriteCustomIndex, profile);
                }
            }
        }

        return FavoriteModeResult.simple();
    }

    public void createProfile(ProfileCreationRequest request, ProfileOperationCallback callback) {
        executorService.execute(() -> {
            try {
                CustomProfile newProfile = new CustomProfile(request.getName());
                newProfile.setIconResId(request.getIconResId());

                for (Map.Entry<String, String> entry : request.getTranslations().entrySet()) {
                    newProfile.setTranslation(entry.getKey(), entry.getValue());
                }

                profileManager.addProfile(newProfile);
                int newIndex = profileManager.getProfiles().size() - 1;

                ProfileOperationResult result = ProfileOperationResult.success(
                        "Profil erfolgreich erstellt", newProfile, newIndex);

                mainHandler.post(() -> callback.onComplete(result));

            } catch (Exception e) {
                ProfileOperationResult result = ProfileOperationResult.error(
                        "Fehler beim Erstellen des Profils", e);
                mainHandler.post(() -> callback.onComplete(result));
            }
        });
    }

    public void deleteProfile(int profileIndex, ProfileOperationCallback callback) {
        executorService.execute(() -> {
            try {
                if (profileIndex < 0 || profileIndex >= profileManager.getProfiles().size()) {
                    throw new IllegalArgumentException("Ungültiger Profil-Index");
                }

                CustomProfile profileToDelete = profileManager.getProfiles().get(profileIndex);
                boolean wasFavorite = profileManager.isFavorite(ProfileManager.MODE_CUSTOM, profileIndex);

                profileManager.setCurrentProfileIndex(profileIndex);
                profileManager.deleteCurrentProfile();

                ProfileDeletionResult deletionResult = new ProfileDeletionResult(
                        profileToDelete, wasFavorite, !profileManager.hasProfiles());

                ProfileOperationResult result = ProfileOperationResult.success(
                        "Profil erfolgreich gelöscht", null, -1, deletionResult);

                mainHandler.post(() -> callback.onComplete(result));

            } catch (Exception e) {
                ProfileOperationResult result = ProfileOperationResult.error(
                        "Fehler beim Löschen des Profils", e);
                mainHandler.post(() -> callback.onComplete(result));
            }
        });
    }

    public void updateProfile(int profileIndex, CustomProfile updatedProfile, ProfileOperationCallback callback) {
        executorService.execute(() -> {
            try {
                if (profileIndex < 0 || profileIndex >= profileManager.getProfiles().size()) {
                    throw new IllegalArgumentException("Ungültiger Profil-Index");
                }

                profileManager.setCurrentProfileIndex(profileIndex);
                profileManager.updateCurrentProfile(updatedProfile);

                ProfileOperationResult result = ProfileOperationResult.success(
                        "Profil erfolgreich aktualisiert", updatedProfile, profileIndex);

                mainHandler.post(() -> callback.onComplete(result));

            } catch (Exception e) {
                ProfileOperationResult result = ProfileOperationResult.error(
                        "Fehler beim Aktualisieren des Profils", e);
                mainHandler.post(() -> callback.onComplete(result));
            }
        });
    }

    public void toggleFavorite(int mode, int customIndex, FavoriteToggleCallback callback) {
        executorService.execute(() -> {
            try {
                boolean wasAlreadyFavorite = profileManager.isFavorite(mode, customIndex);
                profileManager.toggleFavorite(mode, customIndex);
                boolean isNowFavorite = profileManager.isFavorite(mode, customIndex);

                FavoriteToggleResult result = new FavoriteToggleResult(
                        mode, customIndex, wasAlreadyFavorite, isNowFavorite);

                mainHandler.post(() -> callback.onToggleComplete(result));

            } catch (Exception e) {
                mainHandler.post(() -> callback.onError(e));
            }
        });
    }

    // Synchrone Getter für UI-Updates
    public List<CustomProfile> getProfiles() {
        return profileManager.getProfiles();
    }

    public boolean hasProfiles() {
        return profileManager.hasProfiles();
    }

    public CustomProfile getCurrentProfile() {
        return profileManager.getCurrentProfile();
    }

    public int getCurrentProfileIndex() {
        return profileManager.getCurrentProfileIndex();
    }

    public void setCurrentProfileIndex(int index) {
        profileManager.setCurrentProfileIndex(index);
    }

    // Getter für ProfileManager für Legacy-Kompatibilität
    public ProfileManager getProfileManager() {
        return profileManager;
    }

    public void cleanup() {
        if (executorService != null && !executorService.isShutdown()) {
            executorService.shutdown();
        }
    }

    // Data Classes
    public static class FavoriteModeResult {
        private final int mode;
        private final int customIndex;
        private final CustomProfile customProfile;

        private FavoriteModeResult(int mode, int customIndex, CustomProfile customProfile) {
            this.mode = mode;
            this.customIndex = customIndex;
            this.customProfile = customProfile;
        }

        public static FavoriteModeResult simple() {
            return new FavoriteModeResult(ProfileManager.MODE_SIMPLE, -1, null);
        }

        public static FavoriteModeResult extended() {
            return new FavoriteModeResult(ProfileManager.MODE_EXTENDED, -1, null);
        }

        public static FavoriteModeResult custom(int index, CustomProfile profile) {
            return new FavoriteModeResult(ProfileManager.MODE_CUSTOM, index, profile);
        }

        public int getMode() { return mode; }
        public int getCustomIndex() { return customIndex; }
        public CustomProfile getCustomProfile() { return customProfile; }
    }

    public static class ProfileCreationRequest {
        private final String name;
        private final int iconResId;
        private final Map<String, String> translations;

        public ProfileCreationRequest(String name, int iconResId, Map<String, String> translations) {
            this.name = name;
            this.iconResId = iconResId;
            this.translations = translations;
        }

        public String getName() { return name; }
        public int getIconResId() { return iconResId; }
        public Map<String, String> getTranslations() { return translations; }
    }

    public static class ProfileDeletionResult {
        private final CustomProfile deletedProfile;
        private final boolean wasFavorite;
        private final boolean wasLastProfile;

        public ProfileDeletionResult(CustomProfile deletedProfile, boolean wasFavorite, boolean wasLastProfile) {
            this.deletedProfile = deletedProfile;
            this.wasFavorite = wasFavorite;
            this.wasLastProfile = wasLastProfile;
        }

        public CustomProfile getDeletedProfile() { return deletedProfile; }
        public boolean wasFavorite() { return wasFavorite; }
        public boolean wasLastProfile() { return wasLastProfile; }
    }

    public static class FavoriteToggleResult {
        private final int mode;
        private final int customIndex;
        private final boolean wasAlreadyFavorite;
        private final boolean isNowFavorite;

        public FavoriteToggleResult(int mode, int customIndex, boolean wasAlreadyFavorite, boolean isNowFavorite) {
            this.mode = mode;
            this.customIndex = customIndex;
            this.wasAlreadyFavorite = wasAlreadyFavorite;
            this.isNowFavorite = isNowFavorite;
        }

        public int getMode() { return mode; }
        public int getCustomIndex() { return customIndex; }
        public boolean wasAlreadyFavorite() { return wasAlreadyFavorite; }
        public boolean isNowFavorite() { return isNowFavorite; }
    }

    // Callback Interfaces
    public interface FavoriteModeCallback {
        void onSuccess(FavoriteModeResult result);
        void onError(Exception e);
    }

    public interface ProfileOperationCallback {
        void onComplete(ProfileOperationResult result);
    }

    public interface FavoriteToggleCallback {
        void onToggleComplete(FavoriteToggleResult result);
        void onError(Exception e);
    }
}

File: 01_CODE/app/src/main/java/com/beigel/leetSpeak_Generator/ProfileRepository.kt
package com.beigel.leetSpeak_Generator

import android.content.Context
import kotlinx.coroutines.flow.*

/**
 * Modern repository implementing reactive data access with Kotlin Flows
 * Provides clean API for profile management with proper error handling
 */
class ProfileRepository(context: Context) {

    private val leetManager = LeetManager(context)

    // Expose reactive streams from ProfileManager
    val profiles: StateFlow<List<CustomLeet>> = leetManager.profiles
    val currentProfile: StateFlow<CustomLeet?> = leetManager.currentProfile
    val currentProfileIndex: StateFlow<Int> = leetManager.currentProfileIndex
    val hasProfiles: StateFlow<Boolean> = leetManager.hasProfiles
    val favoriteIndex: StateFlow<Int> = leetManager.favoriteIndex

    /**
     * Creates a new profile with the given configuration
     */
    suspend fun createProfile(request: ProfileCreationRequest): ErrorHandler.Result<ProfileCreationResult> =
        ErrorHandler.safeExecute(errorMessage = "Failed to create profile") {
            val profile = CustomLeet(request.name, request.iconResId)
            profile.setTranslations(request.translations)

            val indexResult = leetManager.addProfile(profile)
            when (indexResult) {
                is ErrorHandler.Result.Success -> {
                    ProfileCreationResult(profile, indexResult.data, true, "Profile created successfully")
                }
                is ErrorHandler.Result.Error -> {
                    throw indexResult.exception
                }
            }
        }

    /**
     * Updates an existing profile
     */
    suspend fun updateProfile(index: Int, profile: CustomLeet): ErrorHandler.Result<ProfileUpdateResult> =
        ErrorHandler.safeExecute(errorMessage = "Failed to update profile") {
            leetManager.updateProfile(index, profile).getOrNull()
            ProfileUpdateResult(profile, index, true, "Profile updated successfully")
        }

    /**
     * Deletes a profile at the given index
     */
    suspend fun deleteProfile(index: Int): ErrorHandler.Result<ProfileDeletionResult> =
        ErrorHandler.safeExecute(errorMessage = "Failed to delete profile") {
            val result = leetManager.deleteProfile(index)
            when (result) {
                is ErrorHandler.Result.Success -> {
                    ProfileDeletionResult(
                        deletedProfile = result.data.deletedProfile,
                        wasFavorite = result.data.wasFavorite,
                        wasLastProfile = result.data.wasLastProfile,
                        success = true,
                        message = "Profile deleted successfully"
                    )
                }
                is ErrorHandler.Result.Error -> throw result.exception
            }
        }

    /**
     * Sets the current profile index
     */
    suspend fun setCurrentProfileIndex(index: Int): ErrorHandler.Result<Unit> =
        leetManager.setCurrentProfileIndex(index)

    /**
     * Toggles favorite status for a mode
     */
    suspend fun toggleFavorite(mode: Int, customIndex: Int = 0): ErrorHandler.Result<FavoriteToggleResult> =
        ErrorHandler.safeExecute(errorMessage = "Failed to toggle favorite") {
            val wasAlreadyFavorite = leetManager.isFavorite(mode, customIndex)
            val toggleResult = leetManager.toggleFavorite(mode, customIndex)

            when (toggleResult) {
                is ErrorHandler.Result.Success -> {
                    FavoriteToggleResult(
                        mode = mode,
                        customIndex = customIndex,
                        wasAlreadyFavorite = wasAlreadyFavorite,
                        isNowFavorite = toggleResult.data,
                        success = true
                    )
                }
                is ErrorHandler.Result.Error -> throw toggleResult.exception
            }
        }

    /**
     * Loads the favorite mode asynchronously
     */
    /**
     * Loads the favorite mode asynchronously
     */
    suspend fun loadFavoriteMode(): ErrorHandler.Result<FavoriteModeResult> =
        ErrorHandler.safeExecute(errorMessage = "Failed to load favorite mode") {
            val favoriteInfo = leetManager.getFavoriteMode()

            when {
                favoriteInfo == null -> FavoriteModeResult.simple()
                favoriteInfo.mode == LeetManager.MODE_SIMPLE -> FavoriteModeResult.simple()
                favoriteInfo.mode == LeetManager.MODE_EXTENDED -> FavoriteModeResult.extended()
                favoriteInfo.mode == LeetManager.MODE_CUSTOM -> {
                    if (favoriteInfo.customLeet != null) {
                        FavoriteModeResult.custom(favoriteInfo.customIndex, favoriteInfo.customLeet)
                    } else {
                        FavoriteModeResult.simple() // Fallback
                    }
                }
                else -> FavoriteModeResult.simple() // Fallback für unbekannte Modi
            }
        }

    /**
     * Creates a profile with simple defaults
     */
    suspend fun createProfileWithSimpleDefaults(name: String, iconResId: Int = R.drawable.ic_custom_mode): ErrorHandler.Result<CustomLeet> =
        leetManager.createProfileWithSimpleDefaults(name, iconResId)

    /**
     * Creates a profile with extended defaults
     */
    suspend fun createProfileWithExtendedDefaults(name: String, iconResId: Int = R.drawable.ic_custom_mode): ErrorHandler.Result<CustomLeet> =
        leetManager.createProfileWithExtendedDefaults(name, iconResId)

    /**
     * Gets all leet options for UI display
     */
    fun getLeetOptions(): Flow<List<LeetOption>> = combine(
        profiles,
        currentProfileIndex,
        favoriteIndex
    ) { profiles, currentIndex, favoriteIndex ->
        buildList {
            // Add Simple Leet
            add(LeetOption.createSimple(
                isSelected = currentIndex == -1, // No custom profile selected
                isFavorite = favoriteIndex == LeetManager.FAV_SIMPLE
            ))

            // Add Extended Leet
            add(LeetOption.createExtended(
                isSelected = currentIndex == -2, // Extended mode indicator
                isFavorite = favoriteIndex == LeetManager.FAV_EXTENDED
            ))

            // Add Custom Leets
            profiles.forEachIndexed { index, profile ->
                add(LeetOption.createCustom(
                    profile = profile,
                    customIndex = index,
                    isSelected = currentIndex == index,
                    isFavorite = favoriteIndex == index
                ))
            }
        }
    }

    /**
     * Gets only favorite leet options
     */
    fun getFavoriteLeetOptions(): Flow<List<LeetOption>> =
        getLeetOptions().map { options ->
            options.filter { it.isFavorite }
        }

    /**
     * Checks if a mode is favorite
     */
    fun isFavorite(mode: Int, customIndex: Int = 0): Boolean =
        leetManager.isFavorite(mode, customIndex)

    /**
     * Gets the current profile synchronously (for compatibility)
     */
    fun getCurrentProfile(): CustomLeet? = currentProfile.value

    /**
     * Gets the current profile index synchronously
     */
    fun getCurrentProfileIndex(): Int = currentProfileIndex.value

    /**
     * Checks if profiles exist
     */
    fun hasProfiles(): Boolean = hasProfiles.value

    /**
     * Gets all profiles synchronously
     */
    fun getProfiles(): List<CustomLeet> = profiles.value

    /**
     * Cleanup resources
     */
    fun cleanup() {
        leetManager.cleanup()
    }

    // Legacy compatibility for gradual migration
    fun getProfileManager(): LeetManager = leetManager

    /**
     * Data class for profile creation request
     */
    data class ProfileCreationRequest(
        val name: String,
        val iconResId: Int,
        val translations: Map<String, String>
    )

    /**
     * Result classes for operations
     */
    data class ProfileCreationResult(
        val profile: CustomLeet,
        val index: Int,
        val success: Boolean,
        val message: String
    )

    data class ProfileUpdateResult(
        val profile: CustomLeet,
        val index: Int,
        val success: Boolean,
        val message: String
    )

    data class ProfileDeletionResult(
        val deletedProfile: CustomLeet,
        val wasFavorite: Boolean,
        val wasLastProfile: Boolean,
        val success: Boolean,
        val message: String
    )

    data class FavoriteToggleResult(
        val mode: Int,
        val customIndex: Int,
        val wasAlreadyFavorite: Boolean,
        val isNowFavorite: Boolean,
        val success: Boolean
    )

    /**
     * Favorite mode result sealed class
     */
    sealed class FavoriteModeResult {
        abstract val mode: Int

        data class Simple(override val mode: Int = LeetManager.MODE_SIMPLE) : FavoriteModeResult()
        data class Extended(override val mode: Int = LeetManager.MODE_EXTENDED) : FavoriteModeResult()
        data class Custom(
            override val mode: Int = LeetManager.MODE_CUSTOM,
            val customIndex: Int,
            val profile: CustomLeet
        ) : FavoriteModeResult()

        companion object {
            fun simple() = Simple()
            fun extended() = Extended()
            fun custom(index: Int, profile: CustomLeet) = Custom(customIndex = index, profile = profile)
        }
    }
}